src/main/java/edu/unc/lib/boxc/migration/cdm/services/SourceFilesToRemoteService.java

```diff
@@ -31,58 +38,90 @@ public class SourceFilesToRemoteService {
+            var pathsList = sourceMappings.getMappings().stream()
+                    .map(SourceFilesInfo.SourceFileMapping::getFirstSourcePath)
+                    .collect(Collectors.toList());
+            var pathsDeque = new ConcurrentLinkedDeque<>(pathsList);
+            // Create the parent path structure before we start transfers
+            var parentPaths = listMostSpecificParents(pathsList);
+            createParentPaths(parentPaths, destinationBasePath);
+
+            // Capture futures for transfer threads so we can receive errors
+            var futures = new ArrayList<Future<?>>();
+            for (int i = 0; i < concurrentTransfers; i++) {
+                futures.add(executor.submit(createTransferTask(pathsDeque, destinationBasePath)));
+            }
+            for (Future<?> future : futures) {
+                try {
+                    future.get();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                } catch (ExecutionException e) {
+                    throw new RuntimeException(e.getCause());
+                }
+            }
+        } finally {
+            executor.shutdown();
         }
+    }
 
-        // Wait for all threads to finish
-        threads.forEach(t -> {
-            try {
-                t.join();
-            } catch (InterruptedException e) {
-                throw new RuntimeException("Thread interrupted", e);
+    /**
+     * @param paths list of paths to get parent directories from
+     * @return list of parent directories, deduplicated so that if a directory is a parent of another
+     *  then the parent will be removed. So if we have paths:
+     *      * /a/b
+     *      * /a/b/c
+     *      * /d/e/f
+     * then just the paths /a/b/c and /d/e/f would be returned
+     */
+    private List<Path> listMostSpecificParents(List<Path> paths) {
+        var parents = paths.stream().map(Path::getParent)
+                                    .sorted()
+                                    .collect(Collectors.toList());
+        List<Path> result = new ArrayList<>();
+        for (int i = 0; i < parents.size(); i++) {
+            Path current = parents.get(i);
+            boolean isParent = false;
+            // Check if the current path is a parent of any subsequent path
+            for (int j = i + 1; j < parents.size(); j++) {
+                if (parents.get(j).startsWith(current)) {
+                    isParent = true;
+                    break;
+                }
+            }
+            // If not a parent, add to result
+            if (!isParent) {
+                result.add(current);
+            }
+        }
+        return result;
+    }
+
+    private void createParentPaths(List<Path> paths, Path destinationBasePath) {
+        sshClientService.executeSshBlock((sshClient) -> {
+            for (Path path : paths) {
+                var pathRelative = path.toAbsolutePath().toString().substring(1);
```

Comment:
what is the substring doing here? just taking off the first character?

Reply:
Yes, it's removing the first character, so "/path/to/file" becomes "path/to/file" and then we can evaluate it as a relative path against the base path. If we evaluated "/path/to/file" against "/another/place" we'd just end up with "/path/to/file" since its absolute, but we want "another/place/path/to/file"